File: medium/1143_LongestCommonSubsequence.ts
// 1143. Longest Common Subsequence
// Medium
// https://leetcode.com/problems/longest-common-subsequence

function longestCommonSubsequence(text1: string, text2: string): number {
    let n = text1.length;
    let m = text2.length;
    let dp = [];
    for (let i = 0; i <= n; i++) {
        dp.push(new Array(m + 1).fill(0));
    }
    for (let i = 1; i <= n; i++) {
        for (let j = 1; j <= m; j++) {
            if (text1[i - 1] === text2[j - 1]) {
                dp[i][j] = 1 + dp[i - 1][j - 1];
            } else {
                dp[i][j] = Math.max(dp[i][j - 1], dp[i - 1][j]);
            }
        }
    }

    return dp[n][m];
};

File: medium/1143_LongestCommonSubsequence.py
# 1143. Longest Common Subsequence
# medium
# https://leetcode.com/problems/longest-common-subsequence

class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        n1, n2 = len(text1), len(text2)
        dp = [[0 for _ in range(n2 + 1)] for _ in range(n1 + 1)]
        for l1 in range(n1):
            for l2 in range(n2):
                i, j = l1 + 1, l2 + 1
                if text1[l1] == text2[l2]:
                    dp[i][j] = dp[i - 1][j - 1] + 1
                else:
                    dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
        return dp[n1][n2]
        
'''
dp:
   abc
  0000 
k 0000
l 0000
b 0011
f 0011
c 0012

'''